baseclasses/utils/solverHistory.py

```diff
@@ -0,0 +1,354 @@
+        # Test values are used to check how wide the printed formatted values of a variable will be, so that we can make
+        # sure the columns in the iteration printout are wide enough
+        self.testValues: Dict[Type, Any] = {}
+        # float
+        self.defaultFormat[float] = "{: 17.11e}"
+        self.testValues[float] = 0.0
+        # int
+        self.defaultFormat[int] = "{: 5d}"
+        self.testValues[int] = 0
+        # str
+        self.defaultFormat[str] = "{:^10}"
+        self.testValues[str] = "a"
+        # other
+        self.DEFAULT_OTHER_FORMAT: str = "{}"
+        self.DEFAULT_OTHER_VALUE: list = [2, "b", 3.0]
+
+        # Add fields for the iteration number and time, the only two variables that are always stored
+        self.addVariable("Iter", varType=int, printVar=True)
+        self.addVariable("Time", varType=float, printVar=True, printFormat="{:9.3e}")
+
+    def reset(self, clearMetadata: bool = False) -> None:
+        """Reset the history"""
+
+        # Reset iteration counter
+        self.iter = 0
+
+        # Reset solve start time
+        self.startTime = -1.0
+
+        # Clear recorded data
+        for varName in self.data:
+            self.data[varName] = []
+
+        # Clear metadata if required
+        if clearMetadata:
+            self.metadata = {}
+
+    def addMetadata(self, name: str, data: Any) -> None:
+        """Add a piece of metadata to the history
+
+        The metadata attribute is simply a dictionary that can be used to store arbitrary information related to the
+        solution being recorded, e.g solver options
+
+        Parameters
+        ----------
+        name : str
+            Item name/key
+        data : Any
+            Item to store
+        """
+        self.metadata[name] = data
+
+    def addVariable(
+        self,
+        name: str,
+        varType: Type,
+        printVar: bool = False,
+        printFormat: Optional[str] = None,
+    ) -> None:
+        """Define a new field to be stored in the history.
+
+        Parameters
+        ----------
+        name : str
+            Variable name
+        varType : Type
+            Variable type, i.e int, float, str etc
+        printVar : bool, optional
+            Whether to include the variable in the iteration printout, by default False
+        printFormat : str, optional
+            Format string valid for use with the str.format() method (e.g "{:17.11e}" for a float or "{:03d}" for an
+            int), by default None, in which case a default format for the given `varType` is used, not used if
+            `printVar` is False
+        """
+        self.variables[name] = {"print": printVar, "type": varType}
+
+        # Create variable data list
+        self.data[name] = []
+
+        # Only store variable's string format if it's going to be printed
+        if printVar:
+            if printFormat is not None:
+                self.variables[name]["format"] = printFormat
```

Comment:
Maybe check that this is an actual formatting string? This could be anything.

Reply:
This is a few lines below when generating the test string for the column width computation